File: c#/money-problem.Domain/ConversionResult.cs
namespace money_problem.Domain;


public class ConversionResult<T>
{
    private readonly T? failure;

    private readonly Money? success;

    public ConversionResult(Money success)
    {
        this.success = success;
    }

    public ConversionResult(T failure)
    {
        this.failure = failure;
    }

    public bool IsSuccess() => this.success != null;

    public bool IsFailure() => this.failure != null;

    public T GetFailureUnsafe() => this.failure!;

    public Money GetSuccessUnsafe() => this.success!;
}

File: java/src/main/java/money_problem/domain/ConversionResult.java
package money_problem.domain;

public record ConversionResult(Money money, MissingExchangeRatesException missingExchangeRateException) {
    public ConversionResult(Money money) {
        this(money, null);
    }

    public ConversionResult(MissingExchangeRatesException missingExchangeRateException) {
        this(null, missingExchangeRateException);
    }

    public boolean isFailure() {
        return missingExchangeRateException != null;
    }

    public boolean isSuccess() {
        return money != null;
    }

    public Money success() {
        return money();
    }

    public ConversionError failure() {
        return null;
    }
}
